### PR TITLE
Implement http_cache_forever to ActionController

### PR DESCRIPTION
**Pull request by arthurnn**
Add http_cache_forever to ActionController, so we can cache results
forever.
Things like static pages are a good candidate for this type of caching.

This cache only controls caching headers, so it is up to the browser to
cache those requests.

First stab at #18192 

@dhh @rafaelfranca @sgrif thoughts?

### CODE REVIEW

**Comment by dhh** [comment]
Lovely! Need a CHANGELOG entry as well.

**Comment by arthurnn** [comment]
@dhh Changelog :heavy_check_mark:

File: actionpack/lib/action_controller/caching.rb [hunk 1]
```diff
@@ -46,6 +46,20 @@ def cache_configured?
         end
     end
 
+    # Cache or yield the block. The cache is suppose to never expire.
```
Comment (jeremy) [inline actionpack/lib/action_controller/caching.rb]: This needs to communicate why you'd reach for this method and what it does. You have an HTTP response that never changes, so you want browsers and proxies to cache it indefinitely. On the first hit, we serve the response with long-lived cache freshness response headers. Then subsequent requests will always revalidate as fresh.

This method belongs elsewhere, too. This module is concerned with caching within the app (fragments, template dependencies, etc) not HTTP cache freshness.

File: actionpack/lib/action_controller/caching.rb [hunk 2]
```diff
@@ -46,6 +46,20 @@ def cache_configured?
...
+    #
+    # <tt>version</tt> is the version passed as a key for the cache.
+    def http_cache_forever(public: false, version: 'v1')
+      expires_in 100.years, public: public
+
+      yield if stale?(etag: "#{version}-#{request.fullpath}",
+                      last_modified: Time.parse('2011-01-01').utc,
+                      public: public)
+    end
+
```
Comment (jeremy) [inline actionpack/lib/action_controller/caching.rb]: We can use Last-Modified alone. It's a weak validator (implies semantic equivalence) whereas the ETag is a strong validator (implies byte-for-byte equivalence) which may not be true. We can drop the version arg then.

File: actionpack/lib/action_controller/caching.rb [hunk 3]
```diff
@@ -46,6 +46,20 @@ def cache_configured?
...
     end
 
+    # Cache or yield the block. The cache is suppose to never expire.
+    #
+    # <tt>public</tt> By default the Cache-Control header is private, set this to
+    #   +true+ if you want your application to be cachable by other devices (proxy caches).
+    #
+    # <tt>version</tt> is the version passed as a key for the cache.
+    def http_cache_forever(public: false, version: 'v1')
+      expires_in 100.years, public: public
```
Comment (jeremy) [inline actionpack/lib/action_controller/caching.rb]: Do any browsers or proxies have issues parsing dates this far in the future? e.g. http://en.wikipedia.org/wiki/Year_2038_problem

File: actionpack/lib/action_controller/caching.rb [hunk 4]
```diff
@@ -46,6 +46,20 @@ def cache_configured?
         end
     end
 
+    # Cache or yield the block. The cache is suppose to never expire.
+    #
+    # <tt>public</tt> By default the Cache-Control header is private, set this to
+    #   +true+ if you want your application to be cachable by other devices (proxy caches).
```
Comment (jeremy) [inline actionpack/lib/action_controller/caching.rb]: By default, HTTP responses are private, cached only on the user's web browser. To allow proxies to cache the response, set `public: true` to indicate that they can serve the cached response to all users.

File: actionpack/test/controller/caching_test.rb [hunk 5]
```diff
@@ -349,3 +349,27 @@ def test_view_cache_dependencies_are_listed_in_declaration_order
...
+    assert_equal "max-age=#{100.years.to_i}, public", @response.headers["Cache-Control"]
+    assert_not_nil @response.etag
+  end
+
+  def test_cache_with_private
+    get :cache_me_forever
+    assert_equal "max-age=#{100.years.to_i}, private", @response.headers["Cache-Control"]
+    assert_not_nil @response.etag
+  end
+end
```
Comment (jeremy) [inline actionpack/test/controller/caching_test.rb]: This tests that a request without cache freshness headers results in a response with headers properly set. These should assert 200 OK responses and additionally test that requests with cache freshness headers result in 304 Not Modified responses.

File: actionpack/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,7 @@
+*   Add http_cache_forever to ActionController, so we can cache results that won't expire.
```
Comment (jeremy) [inline actionpack/CHANGELOG.md]: s/ActionController/Action Controller
s/results that won't/responses that never

File: actionpack/lib/action_controller/caching.rb (hunk 1 above)
Comment (arthurnn) [inline actionpack/lib/action_controller/caching.rb]: Maybe should I move this to https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/metal/conditional_get.rb ?

File: actionpack/CHANGELOG.md (hunk 6 above)
Comment (arthurnn) [inline actionpack/CHANGELOG.md]: done

File: actionpack/test/controller/caching_test.rb (hunk 5 above)
Comment (arthurnn) [inline actionpack/test/controller/caching_test.rb]: Done, please check it out the last commit

File: actionpack/lib/action_controller/caching.rb (hunk 4 above)
Comment (arthurnn) [inline actionpack/lib/action_controller/caching.rb]: done

File: actionpack/lib/action_controller/caching.rb (hunk 2 above)
Comment (arthurnn) [inline actionpack/lib/action_controller/caching.rb]: we need the etag too so we can bust the cache adding a new version, see my test case `test_cache_response_code_with_etag` .

**Comment by arthurnn** [comment]
@jeremy I addressed most of your concerns, just leaving the where this should be still for decision. Let me know your thoughts. thanks

File: actionpack/lib/action_controller/caching.rb [hunk 7]
```diff
@@ -46,6 +46,21 @@ def cache_configured?
...
+    # <tt>public</tt> By default, HTTP responses are private, cached only on the
+    #   user's web browser. To allow proxies to cache the response, set +true+ to
+    #   indicate that they can serve the cached response to all users.
+    #
+    # <tt>version</tt> is the version passed as a key for the cache.
+    def http_cache_forever(public: false, version: 'v1')
+      expires_in 100.years, public: public
+
+      yield if stale?(etag: "#{version}-#{request.fullpath}",
+                      last_modified: Time.parse('2011-01-01').utc,
```
Comment (andyatkinson) [inline actionpack/lib/action_controller/caching.rb]: What was the significance of this date (curious)? Thanks.

**Comment by andyatkinson** [comment]
@arthurnn I worked on a PR but then I realized yours was much further along. I fetched yours, ran the tests, and read through them. Looks good! :+1:

File: actionpack/lib/action_controller/caching.rb (hunk 7 above)
Comment (arthurnn) [inline actionpack/lib/action_controller/caching.rb]: this is just a date on the past. But now I wonder if the last_modifed is necessary at all. I guess we could use only the etag, as that one has the version. Not 100% sure tho, as this is the opposite what @jeremy said in the comment above.. lets wait for him on this.

**Comment by dhh** [comment]
@jeremy Can you give this a final look over? Seems like we're just about ready for a merge.

**Comment by arthurnn** [comment]
@jeremy let me know your thoughts, I can fix/rebase/merge this when you fell this is ready!

File: actionpack/lib/action_controller/caching.rb (hunk 2 above)
Comment (jeremy) [inline actionpack/lib/action_controller/caching.rb]: Sounds good. We can just switch to weak ETags when we implement them.

**Comment by jeremy** [comment]
:ship:

**Comment by kaspth** [comment]
@arthurnn just a reminder that you're also working on this and need to rebase :smile:

**Comment by arthurnn** [comment]
merging this today, after breakfast =)

**Comment by kaspth** [comment]
forever :heart:

**Comment by arthurnn** [comment]
:tada: :smile_cat:

**Comment by dhh** [comment]
Boom! :metal:

**Pull request merged**